eventsourcing/src/main/java/org/axonframework/eventsourcing/annotation/AnnotationBasedEventCriteriaResolver.java

```diff
@@ -90,32 +100,108 @@ public AnnotationBasedEventCriteriaResolver(@Nonnull Class<?> entityType) {
+     */
+    private class WrappedEventCriteriaBuilderMethod {
+
+        private final Method method;
+        private final Class<?> identifierType;
+        private final Object[] optionalArgumentSuppliers;
+
+        private WrappedEventCriteriaBuilderMethod(Method method) {
+            if (!EventCriteria.class.isAssignableFrom(method.getReturnType())) {
+                throw new IllegalArgumentException(
+                        "Method annotated with @EventCriteriaBuilder must return an EventCriteria. Violating method: %s".formatted(
+                                ReflectionUtils.toDiscernibleSignature(method)));
+            }
+            if (!Modifier.isStatic(method.getModifiers())) {
+                throw new IllegalArgumentException(
+                        "Method annotated with @EventCriteriaBuilder must be static. Violating method: %s".formatted(
+                                ReflectionUtils.toDiscernibleSignature(method)));
+            }
+            if (method.getParameterCount() == 0) {
+                throw new IllegalArgumentException(
+                        "Method annotated with @EventCriteriaBuilder must have at least one parameter which is the identifier. Violating method: %s".formatted(
+                                ReflectionUtils.toDiscernibleSignature(method)));
+            }
+            this.method = ReflectionUtils.ensureAccessible(method);
+            // Let's determine the identifier and create the argument suppliers
+            this.identifierType = method.getParameterTypes()[0];
+            int optionalParameterCount = method.getParameterCount() - 1;
+            this.optionalArgumentSuppliers = new Object[optionalParameterCount];
+
+            // And other arguments which can come from the configuration
+            for (int i = 0; i < optionalParameterCount; i++) {
+                Class<?> parameterType = method.getParameterTypes()[i + 1];
+
+                // The whole configuration can be passed in
+                if (parameterType == NewConfiguration.class) {
+                    optionalArgumentSuppliers[i] = configuration;
+                    continue;
+                }
+                // Or a specific component of the configuration
+                Optional<?> component = configuration.getOptionalComponent(parameterType);
+                if (component.isEmpty()) {
+                    throw new IllegalArgumentException(
+                            "Method annotated with @EventCriteriaBuilder declared a parameter which is not a component: %s. Violating method: %s".formatted(
+                                    parameterType.getName(),
+                                    ReflectionUtils.toDiscernibleSignature(method)
+                            ));
+                }
+                optionalArgumentSuppliers[i] = component.get();
+            }
+        }
+
+        public Object resolve(Object id) {
```

Comment:
I think public modifiers in this class aren't needed, because the whole class is private.